Java: Numeric Type Promotion (#2224)
cap-js/docs: java/cds-data.md

These types are used for the values of CDS elements with primitive type. In the [Model Reflection API](./reflection-api), they're represented by the enum [CdsBaseType](https://javadoc.io/doc/com.sap.cds/cds4j-api/latest/com/sap/cds/reflect/CdsBaseType.html).

### Numeric Type Determination { #type-determination }

To have a consistent behavior across different databases, the CAP Java runtime applies numeric type determination in arithmetic expressions and numeric standard functions according to the following rules.

::: tip
Use `type(CdsBaseType)` to explicitly set the result type if needed.
:::

#### Arithmetic Expressions { #numeric-promotion }

Arithmetic expressions promote numeric types according to the following precedence:

**Type Precedence (highest to lowest):**

> `DOUBLE` (binary64), `HANA_REAL` (binary32), `DECIMAL` (base 10), `INT64`, `INT32`, `INT16`, `UINT8`

Reviewer: ```suggestion
`DOUBLE`, `HANA_REAL`, `DECIMAL`, `INT64`, `INT32`, `INT16`, `UINT8`
```

- For addition, subtraction, and multiplication, the result type is the one with highest precedence among the operands.
- For division:

>   - If any operand is a binary floating-point type (`DOUBLE`, `HANA_REAL`), the result type is `DOUBLE`.

Reviewer: ```suggestion
  - If any operand is a binary approximate numeric type (`DOUBLE`, `HANA_REAL`), the result type is `DOUBLE`.
```

  - Otherwise, the result type is `DECIMAL`, which provides higher accuracy for decimal fractions.

#### Numeric Standard Functions { #type-determination-functions }

Numeric aggregation and standard functions determine their result type based on the argument types:

- **ceiling(x)**, **floor(x)**, **round(x)**: Return the same type as the input `x`.
- **min(x)**, **max(x)**, **sum(x)**: Aggregate functions return the same type as the argument `x`.
- **average(x)**: Returns `DECIMAL` for exact numeric types and `DOUBLE` for approximate numeric types.

> - **count(x)**, **countdistinct(x)**: Always return `INT64`.

Reviewer: ```suggestion
- **count(x)**, **countdistinct(x)**: return `INT64`.
```

> - **count(x)**, **countdistinct(x)**: return `cds.Int64`.

Reviewer 2: ```suggestion
- **count(x)**, **countdistinct(x)**: Return `cds.Int64`.
```
